### PR TITLE
ext-proc: Support "immediate_response" options for request headers 

### PR DESCRIPTION
**Pull request by gbrail**
This lets ext_proc servers return an immediate HTTP response (such as to indicate an error) in response to a request_headers message.

Signed-off-by: Gregory Brail <gregbrail@google.com>

### CODE REVIEW

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -129,6 +130,23 @@ void Filter::onGrpcClose() {
   }
 }
 
+void Filter::sendImmediateResponse(const ImmediateResponse& response) {
+  const auto status_code = response.has_status() ? response.status().code() : 200;
+  const auto grpc_status =
+      response.has_grpc_status()
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:136]: Is this behavior covered in the tests below?

**Review by htuch — COMMENTED** [review]
Looks good, one question..

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 1 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:136]: Line coverage was OK but I think the coverage of all the cases could be better -- fixed that in the unit test.

File: test/extensions/filters/http/ext_proc/filter_test.cc [hunk 2]
```diff
@@ -265,6 +279,92 @@ TEST_F(HttpFilterTest, PostAndRespondImmediately) {
...
+                           std::function<void(Http::ResponseHeaderMap & headers)> modify_headers,
+                           Unused, Unused) { modify_headers(immediate_response_headers); }));
+  std::unique_ptr<ProcessingResponse> resp1 = std::make_unique<ProcessingResponse>();
+  auto* immediate_response = resp1->mutable_immediate_response();
+  immediate_response->mutable_status()->set_code(envoy::type::v3::StatusCode::Forbidden);
+  immediate_response->mutable_grpc_status()->set_status(999);
+  stream_callbacks_->onReceiveMessage(std::move(resp1));
+  EXPECT_FALSE(stream_close_sent_);
+
+  Http::TestResponseHeaderMapImpl expected_response_headers{};
```
Comment (htuch) [inline test/extensions/filters/http/ext_proc/filter_test.cc:350]: Why do we always get empty response headers?

File: test/extensions/filters/http/ext_proc/filter_test.cc [hunk 3]
```diff
@@ -265,6 +279,92 @@ TEST_F(HttpFilterTest, PostAndRespondImmediately) {
...
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_->encodeHeaders(response_headers_, true));
+  data_.add("bar");
+  EXPECT_EQ(FilterDataStatus::Continue, filter_->encodeData(data_, false));
+  EXPECT_EQ(FilterDataStatus::Continue, filter_->encodeData(data_, true));
+  EXPECT_EQ(FilterTrailersStatus::Continue, filter_->encodeTrailers(response_trailers_));
+  filter_->onDestroy();
+  EXPECT_TRUE(stream_close_sent_);
+}
+
+TEST_F(HttpFilterTest, RespondImmediatelyGrpcError) {
```
Comment (htuch) [inline test/extensions/filters/http/ext_proc/filter_test.cc:324]: Nit: for these more involved tests, can you add `//` one liners above each explaining what is being tested? Thanks.

**Review by htuch — COMMENTED** [review]
Yep, those were the tests I was looking for. Couple of nits and LGTM.

File: test/extensions/filters/http/ext_proc/filter_test.cc (hunk 2 above)
Comment (gbrail) [reply test/extensions/filters/http/ext_proc/filter_test.cc:367]: This particular test is a unit test that drives the filter directly, so what we're testing is whether the instructions in the "immediate response" message successfully modified the headers. The "PostAndRespondImmediately" test above checks for this. I chose not to try and set headers in all the other tests of that feature just to reduce the amount of cutting and pasting.

File: test/extensions/filters/http/ext_proc/filter_test.cc (hunk 3 above)
Comment (gbrail) [reply test/extensions/filters/http/ext_proc/filter_test.cc:341]: I added those in a new commit.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**